### DISCUSSION TITLE
Stores lag behind routes in SSR: intentional?

### DISCUSSION THREAD

**Discussion by baukevanderlaan** [sveltejs/kit#3209] [Q&A] [answered]
I’m bitten several times by the serverside rendered lag in stores that happens when a store is set on a page, but is read by a component that's higher up in `__layout`. The only workaround I found, is moving the component that reads the store to a place lower than `<slot/>` but that isn’t always an option. This leads me to think that this is intentional, after all the store can't be read before it's set.

I made a minimal reproduction to show the issue:

1. Get this reproduction https://github.com/baukevanderlaan/sveltekit-store-reproduction
2. Disable Javascript in your browser (or view source when inspecting the pages)
3. Build and preview the reproduction
5. Click around the navigation, notice the store value is empty on first pageview, and then lags behind when navigating to another page. Because the store state lag is on the server, this also occurs when you open the preview in two tabs, navigate in one tab, then refresh the page in the other tab.
6. Enable Javascript and observe that hydration 'fixes' the lag.
7. Move the `<DisplayStore />` below `<slot/>` in `src/routes/__layout` and observe that this fixes the lag as well.

Is this like it should work? If so: what is the recommended way to deal with this problem? I'm using stores for example to set `<meta>`-tags, a language switcher, breadcrumbs etc. All based on the page content, but higher up in the component hierarchy.

**Comment by benmccann** [comment, marked as answer]
See https://github.com/sveltejs/kit/issues/269#issuecomment-892211523 for more details about this. I think this limitation could be better documented. Probably the most immediate improvement here would be built-in support for metadata in SvelteKit, which is something we'd like to do

**Reply by baukevanderlaan** [reply]
Thanks for your reply @benmccann!

I read through that issue and if I understand it correctly, the problem with SSR store lag lies in the method of building the HTML string that doesn't wait for all stores to settle. I understand the alternative would take long to render, but to me the fact that I now need to take into account the component hierarchy in order to have an up to date store is a bit of a bummer as it diminishes the convenience that Svelte stores offered in not having to think about where a component is positioned.

On the surface, building in a ‘special’ store for metadata feels a bit like sidestepping the problem; I feel like this is broader than metadata alone and I imagine that there will always be cases in which data (other than the contents of `<head>`) needs to be passed from lower level components to higher level components. Would it be appreciated if I voice that concern in #269, or have you already decided on this?

Thanks again!

**Reply by benmccann** [reply]
The metadata would not be just for `<meta>` tags in `<head>`, but for passing any data upwards. Metadata might not be the best name. We haven't really thought too hard about the name yet

**Comment by baukevanderlaan** [comment]
I’ve made a deployment of the reproduction for easier viewing: https://sveltekit-store-ssr.vercel.app/

**Comment by Rich-Harris** [comment]
The solution we've come up with for this class of problem is https://github.com/sveltejs/kit/pull/3252. Anything added to [`stuff`](https://kit.svelte.dev/docs#loading-output-stuff) by a layout or page component will be available at render time as `$page.stuff`
